general/handle.hpp

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2010-2024, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_HANDLE_HPP
+#define MFEM_HANDLE_HPP
+
+#include "../config/config.hpp"
+#include <memory>
+
+namespace mfem
+{
+
+/// @brief A smart pointer class that may represent either shared ownership, or
+/// a non-owning borrow.
+///
+/// A Handle may either be owning or non-owning. Non-owning Handle%s point to
+/// externally owned data; it is the responsibility of the user both to ensure
+/// that the data remains valid as long as the Handle is alive and to delete the
+/// pointer when it is no longer needed. Owning Handle%s use <a
+/// href="https://en.cppreference.com/w/cpp/memory/shared_ptr">
+/// std::shared_ptr</a> to implement reference counting. The underlying data
+/// will be valid as long as there is at least one live copy. When the last
+/// Handle is destroyed, the pointer is deleted.
+///
+/// Both types of Handle%s can be copied, moved, stored in standard containers,
+/// etc.
+///
+/// A non-owning Handle may assume ownership over its data, but an owning Handle
+/// cannot release ownership over its data.
+///
+/// It is an invariant of this class that **at most** one of the data members
+/// @a not_owned and @a owned will be non-null.
+template <typename T>
+class Handle
+{
+   /// If this is a non-owning handle, @a not_owned will point to the data.
+   T *not_owned = nullptr;
+
+   /// If this is an owning handle, @a owned will point to the data.
+   std::shared_ptr<T> owned = nullptr;
+
+   /// @brief Types @a Handle<T> and @a %Handle\<U\> are friends to allow
+   /// construction of one from another when @a T and @a U are convertible
+   /// types.
+   template <typename U> friend class Handle;
+
+public:
+   /// Create an empty (null) Handle.
+   Handle() = default;
+
+   /// @brief Create a Handle pointing to @a t.
+   ///
+   /// If @a take_ownership is true, then the Handle assumes ownership over the
+   /// pointer, and it should not be deleted externally. Otherwise, the Handle
+   /// will be non-owning, and it is the user's responsibility to ensure the
+   /// correct lifetime of @a t.
+   Handle(T *t, bool take_ownership)
+   {
+      if (take_ownership) { owned.reset(t); }
+      else { not_owned = t; }
+   }
```

Comment:
What about basing whether it's owned by the type of the argument:
```suggestion
   /// @brief Create a Handle pointing to non-owned @a t.
   ///
   ///  Create an non-owning Handle; it is the user's responsibility to ensure the
   /// correct lifetime of @a t.
   Handle(T *t) : not_owned(t)
   {
   }
   
      /// @brief Create a Handle pointing to non-owned @a t.
   ///
   ///  Create an owning Handle to @a t.
   Handle(::std::unique_ptr<T> t) : owned(t)
   {
   }
```

Reply:
The `Handle(std::unique_ptr<T> t)` overload could be useful (e.g. to create an owning handle directly from the return value of a function that returns `unique_ptr`), but does making it the only mechanism offer a concrete benefit? The type is essentially playing the role of a flag, since the `unique_ptr` is never used as such, it's just immediately released.

Reply:
For me, the meaning of a `T *t` argument is very ambiguous, and I can't know what is intended without reading the comments and/or the implementation. A raw pointer is ambiguous about ownership, and it's ambiguous about size. Is it a pointer to zero or one object, or is it the start of an array of objects. `std::unique_ptr` is clear about ownership transfer, and array or not.
If you always use `std::unique_ptr` for ownership transfer, users can infer that non-smart raw pointers indicates a borrowed reference.

Reply:
I'm not sure the "all ownership transfers happen via `unique_ptr`" rule can be followed closely enough so that the inference that all raw pointers are non-owning borrows can be made reliably (and in any case, non-owning borrows can be modeled with references instead of pointers).

For example, the `shared_ptr::shared_ptr(T*)` and `shared_ptr::reset(T*)` assume ownership of the passed pointer. I'm really not sure that `shared_ptr::shared_ptr(unique_ptr<T>)` and `shared_ptr::reset(unique_ptr<T>)` would be a better interface (it introduces a certain amount of "ceremony" with IMO little benefit in clarity). But since we use the standard library (and many external libraries), we can't reliably assume that all uses of raw pointers are non-owning.

Reply:
I totally agree about using reference arguments instead of pointers when `nullptr` isn't an option.

My impression as a user is that inconsistent pointer use is one of the challenging things about using MFEM. It's pretty common for me to have to read a method's implementation to figure out how the pointer is being used. With other C/C++ libraries that I use, it's not common for me to have to read method implementations to figure things out. This is probably due to a combination of documentation and consistent interface design.

Reply:
I agree completely. Inconsistent pointer usage is a pain point when using MFEM, and users shouldn't have to read the implementation to understand the ownership semantics (but often do). We should work towards addressing both of these issues in the next major release of MFEM.